src/ape/api/projects.py

```diff
@@ -0,0 +1,229 @@
+import json
+import shutil
+from pathlib import Path
+from typing import TYPE_CHECKING, ClassVar, Collection, Dict, Optional
+
+from ethpm_types import Checksum, ContractType, PackageManifest, Source
+from ethpm_types.manifest import PackageName
+from ethpm_types.utils import compute_checksum
+
+from ape.exceptions import ProjectError
+from ape.logging import logger
+from ape.types import DependencyType
+from ape.utils import (
+    abstractmethod,
+    get_all_files_in_directory,
+    get_relative_path,
+    injected_before_use,
+)
+
+if TYPE_CHECKING:
+    from ape.managers.compilers import CompilerManager
+
+
+class ProjectAPI:
+    """
+    An abstract base-class for working with projects. The top-level
+    :class:`ape.managers.project.ProjectManager` implements this class. This class is
+    also a base-class for dependencies (see :class:`~ape.api.projects.DependencyAPI`,
+    such as from Github, that get cached as manifest files.
+
+    This class can also be extended to a plugin for supporting non-ape projects.
+    """
+
+    def __repr__(self):
+        return f"<{self.__class__.__name__}, name={self.name}>"
+
+    path: Path
+    """The project path."""
+
+    compilers: ClassVar["CompilerManager"] = injected_before_use()  # type: ignore
+    """
+    The group of compiler plugins for compiling source files. See
+    :class:`~ape.managers.compilers.CompilerManager` for more information.
+    Call method :meth:`~ape.managers.project.ProjectManager.load_contracts` in this class
+    to more easily compile sources.
+    """
+
+    def __init__(
+        self,
+        path: Path,
+    ) -> None:
+        self.path = Path(path) if isinstance(path, str) else path
+
+    @property
+    @abstractmethod
+    def name(self) -> str:
+        """
+        The name of the project.
+        """
+
+    @property
+    @abstractmethod
+    def cache_folder(self) -> Path:
+        """
+        The folder storing the manifest.
+        """
+
+    @property
+    @abstractmethod
+    def version(self) -> str:
+        """
+        The version of the project.
+        """
+
+    @property
+    def manifest_cachefile(self) -> Path:
```

Comment:
This seems sort of like an internal optimization (to eliminate having to reload every time)